proof/refine/ARM/KHeap_R.thy

```diff
@@ -2534,6 +2573,25 @@ lemma setReply_state_refs_of'[wp]:
    \<lbrace>\<lambda>rv s. P (state_refs_of' s)\<rbrace>"
   by (wp set_reply'.state_refs_of') (simp flip: fun_upd_def)
 
+lemma setReply_replyNexts_replyPrevs[wp]:
+  "\<lbrace>\<lambda>s. P ((replyNexts_of s)(rptr := replyNext_of reply))
+          ((replyPrevs_of s)(rptr := replyPrev reply))\<rbrace>
+   setReply rptr reply
+   \<lbrace>\<lambda>_ s. P (replyNexts_of s) (replyPrevs_of s)\<rbrace>"
+  apply (wpsimp simp: setReply_def updateObject_default_def setObject_def split_def)
+  apply (erule rsubst2[where P=P])
+   apply (clarsimp simp: ext opt_map_def list_refs_of_reply'_def map_set_def projectKO_opt_reply
+                  split: option.splits)+
+  done
+
+lemma updateReply_wp_all:
+  "\<lbrace>\<lambda>s. \<forall>ko. ko_at' ko rptr s \<longrightarrow> P (set_obj' rptr (upd ko) s)\<rbrace>
+   updateReply rptr upd
+   \<lbrace>\<lambda>_. P\<rbrace>"
+  unfolding updateReply_def
+  apply (wpsimp wp: set_reply'.set_wp)
```

Comment:
nitpick: use by in these very short proofs. This is a true nitpick, feel free to ignore.